st3/lsp_utils/npm_client_handler_v2.py

```diff
@@ -0,0 +1,176 @@
+    'settings': {},
+}  # type: ignore
+
+
+def is_node_installed():
+    return shutil.which('node') is not None
+
+
+class ApiWrapper(object):
+    def __init__(self, client):
+        self.__client = client
+
+    def on_notification(self, method: str, handler: Callable) -> None:
+        pass
+        # self.__client.on_notification(method, handler)
+
+    def on_request(self, method: str, handler: Callable) -> None:
+        pass
+        # def on_response(params, request_id):
+        #     handler(params, lambda result: send_response(request_id, result))
+
+        # def send_response(request_id, result):
+        #     self.__client.send_response(Response(request_id, result))
+
+        # self.__client.on_request(method, on_response)
+
+
+class NpmClientHandler(Session):
+    package_name = ''  # type: str
+    server_directory = ''  # type: str
+    server_binary_path = ''  # type: str
+    # Internal
+    __server = None  # type: ServerNpmResource
+    __settings_filename = None
+
+    @classmethod
+    def setup(cls) -> None:
+        if not cls.__server:
+            cls.__settings_filename = '{}.sublime-settings'.format(cls.package_name)
+            cls.__server = ServerNpmResource(cls.package_name, cls.server_directory, cls.server_binary_path)
+            cls.__server.setup()
+
+    @classmethod
+    def cleanup(cls) -> None:
+        cls.__server.cleanup()
+
+    @classmethod
+    def name(cls) -> str:
+        return cls.package_name.lower()
+
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        return False
+
+    @classmethod
+    def install_or_update(cls) -> None:
+        pass
+
+    @classmethod
+    def standard_configuration(cls) -> ClientConfig:
+        cls.setup()
+
+        assert cls.__server
+
+        configuration = {
+            'enabled': True,
+            'command': ['node', cls.__server.binary_path] + cls.get_binary_arguments(),
+        }
+
+        configuration.update(cls._read_configuration())
+        cls.on_client_configuration_ready(configuration)
+        return read_client_config(cls.name(), configuration)
+
+    @classmethod
+    def get_binary_arguments(cls):
+        """
+        Returns a list of extra arguments to append when starting server.
+        """
+        return ['--stdio']
+
+    @classmethod
+    def _read_configuration(cls) -> Dict:
+        settings = {}  # type: Dict
+        loaded_settings = sublime.load_settings(cls.__settings_filename)
+
+        if loaded_settings:
+            migrated = cls._migrate_obsolete_settings(loaded_settings)
+            changed = cls.on_settings_read(loaded_settings)
+            if migrated or changed:
+                sublime.save_settings(cls.__settings_filename)
+
+            for key, default in CLIENT_SETTING_KEYS.items():
+                settings[key] = loaded_settings.get(key, default)
+
+        return settings
+
+    @classmethod
+    def _migrate_obsolete_settings(cls, settings: sublime.Settings):
```

Comment:
What is the difference between `_migrate_obsolete_settings`and  `on_settings_read`? 

Both can be used to migrate them to new schema, 
but `on_settings_read` can be used to change other settings while `_migrate_obsolete_settings` (probably) should not?

Can you clarify it for me? :)

Reply:
`_migrate_obsolete_settings` is an internal function that always runs and can't be overridden. And also can be removed soon because we can assume that all configurations were already migrated.

`on_settings_read` is the one that subclasses can override and do other kind of migration.